src/merge_operator.rs

```diff
@@ -451,4 +454,40 @@ mod test {
         }
         assert!(DB::destroy(&opts, path).is_ok());
     }
+
+    #[test]
+    fn failed_merge_test() {
+        fn test_failing_merge(
+            _key: &[u8],
+            _val: Option<&[u8]>,
+            _operands: &mut MergeOperands,
+        ) -> Option<Vec<u8>> {
+            None
+        }
+        use crate::{Options, DB};
+
+        let path = "_rust_rocksdb_failed_merge_test";
+        let mut opts = Options::default();
+        opts.create_if_missing(true);
+        opts.set_merge_operator("test operator", test_failing_merge, None);
+        {
+            let db = DB::open(&opts, path).expect("open with a merge operator");
+            db.put(b"key", b"value").expect("put_ok");
+            let res = db.merge(b"key", b"new value");
+            match res.and_then(|_e| db.get(b"key")) {
+                Ok(val) => panic!("expected merge failure to propagate, got: {:?}", val),
+                Err(e) => {
+                    assert!(e.into_string().contains("Could not perform merge."));
+                }
+            }
+            // DB::destroy detects the corruption and returns an error
+            // without deleting all of the files, so we ignore its
+            // return value and then call `fs::remove_dir_all` to be
+            // sure we've cleaned up.
+            match DB::destroy(&opts, path) {
+                _ => (),
+            };
+            fs::remove_dir_all(path).expect("deleting test database");
```

Comment:
```rust
if DB::destroy(&opts, path).is_err() {
    fs::remove_dir_all(path).expect("deleting test database");
}
```
???

Reply:
As far as I can see, the reason of the failure is a locked file, so I doubt it will solve the issue, but we can try. :upside_down_face:

Reply:
Time to move all tests to the `tests` module and use `DBPath` everywhere 😄

Reply:
`DBPath` would also fail here, for the same a `DB::destroy` alone doesn't work, but we could amend it to handle this case somehow. Do you want me to make that change?

Reply:
Ideally, we need to find out why this test is having issues with deleting files, but as a quick solution personally I would use `TempDir`.

Reply:
Pushed a fix here! I realized now that I've been confused why `DB::destroy` was failing; I'd been blaming on the corrupted database, but the real problem was just that I was calling it while `db` was still alive, and so it couldn't lock the database. So now we use a tempfile, and also `DB::destroy().expect()`. Still passing locally.

Reply:
Cool. :+1: 

Though as the problem is resolved I would prefer to use `DBPath` instead of tempdir + manual call to `DB::destroy()`.

Reply:
I moved all the merge_operator tests to `tests/` and to use `DBPath`; let me know if you'd rather I roll that back and send another PR.